core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2200,4 +2197,222 @@ public void testDocStats() throws IOException {
+        if (randomBoolean()) {
+            engine.index(operation);
+            assertNotNull(operation.getTranslogLocation());
+            engine.index(retry);
+            assertNotNull(retry.getTranslogLocation());
+            assertTrue(retry.getTranslogLocation().compareTo(operation.getTranslogLocation()) > 0);
+        } else {
+            engine.index(retry);
+            assertNotNull(retry.getTranslogLocation());
+            engine.index(operation);
+            assertNotNull(retry.getTranslogLocation());
+            assertTrue(retry.getTranslogLocation().compareTo(operation.getTranslogLocation()) < 0);
+        }
+
+        engine.refresh("test");
+        try (Engine.Searcher searcher = engine.acquireSearcher("test")) {
+            TopDocs topDocs = searcher.searcher().search(new MatchAllDocsQuery(), 10);
+            assertEquals(1, topDocs.totalHits);
+        }
+    }
+
+
+    public void testRetryWithAutogeneratedIdWorksAndNoDuplicateDocs() throws IOException {
+
+        final ParsedDocument doc = testParsedDocument("1", "1", "test", null, 100, -1, testDocumentWithTextField(), new BytesArray("{}".getBytes(Charset.defaultCharset())), null);
+        boolean canHaveDuplicates = false;
+        boolean autoGeneratedId = true;
+
+        Engine.Index index = new Engine.Index(newUid("1"), doc, Versions.MATCH_ANY, VersionType.INTERNAL, PRIMARY, System.nanoTime(), autoGeneratedId, canHaveDuplicates);
+        engine.index(index);
+        assertThat(index.version(), equalTo(1L));
+
+        index = new Engine.Index(newUid("1"), doc, index.version(), index.versionType().versionTypeForReplicationAndRecovery(), REPLICA, System.nanoTime(), autoGeneratedId, canHaveDuplicates);
+        replicaEngine.index(index);
+        assertThat(index.version(), equalTo(1L));
+
+        canHaveDuplicates = true;
+        index = new Engine.Index(newUid("1"), doc, Versions.MATCH_ANY, VersionType.INTERNAL, PRIMARY, System.nanoTime(), autoGeneratedId, canHaveDuplicates);
+        engine.index(index);
+        assertThat(index.version(), equalTo(2L));
+        engine.refresh("test");
+        try (Engine.Searcher searcher = engine.acquireSearcher("test")) {
+            TopDocs topDocs = searcher.searcher().search(new MatchAllDocsQuery(), 10);
+            assertEquals(1, topDocs.totalHits);
+        }
+
+        index = new Engine.Index(newUid("1"), doc, index.version(), index.versionType().versionTypeForReplicationAndRecovery(), REPLICA, System.nanoTime(), autoGeneratedId, canHaveDuplicates);
+        replicaEngine.index(index);
+        replicaEngine.refresh("test");
+        try (Engine.Searcher searcher = replicaEngine.acquireSearcher("test")) {
+            TopDocs topDocs = searcher.searcher().search(new MatchAllDocsQuery(), 10);
+            assertEquals(1, topDocs.totalHits);
+        }
+    }
+
+    public void testRetryWithAutogeneratedIdsAndWrongOrderWorksAndNoDuplicateDocs() throws IOException {
```

Comment:
do we have a  variant where the original request arrives second at the primary but first at the replica (and the reverse)?